src/main/java/graphql/execution/ExecutionStrategy.java

```diff
@@ -120,25 +122,29 @@ protected ExecutionResult completeValue(ExecutionContext executionContext, Graph
         return executionContext.getQueryStrategy().execute(executionContext, resolvedType, result, subFields);
     }
 
-    private ExecutionResult completeValueForList(ExecutionContext executionContext, GraphQLList fieldType, List<Field> fields, Object result) {
+    private ExecutionResult completeValueForList(ExecutionContext executionContext, GraphQLList fieldType, List<Field> fields, Object result, Map<String, Object> argumentValues) {
         if (result.getClass().isArray()) {
             result = Arrays.asList((Object[]) result);
         }
 
         //noinspection unchecked
-        return completeValueForList(executionContext, fieldType, fields, (Iterable<Object>) result);
+        return completeValueForList(executionContext, fieldType, fields, (Iterable<Object>) result, argumentValues);
     }
 
-    protected GraphQLObjectType resolveType(GraphQLInterfaceType graphQLInterfaceType, Object value) {
-        GraphQLObjectType result = graphQLInterfaceType.getTypeResolver().getType(value);
+    protected GraphQLObjectType resolveType(Field field, GraphQLInterfaceType graphQLInterfaceType,
+                                            Object value, Map<String, Object> argumentValues, GraphQLSchema schema) {
+        TypeResolutionEnvironment env = new TypeResolutionEnvironment(value, argumentValues, field, graphQLInterfaceType, schema);
+        GraphQLObjectType result = graphQLInterfaceType.getTypeResolver().getType(env);
         if (result == null) {
             throw new GraphQLException("could not determine type");
         }
         return result;
     }
 
-    protected GraphQLObjectType resolveType(GraphQLUnionType graphQLUnionType, Object value) {
```

Comment:
I've reimplemented the changes to the `ExecutionStrategy` in the way you suggested and made overloads and deprecations for backward-compatibility. Please see when you get a chance what you think of this version.

**Notes:**
- I've only introduced the `...Parameters` classes where I needed them for this change, not everywhere where they should arguably go. Tell me If you want me refactor more places.
- I made the `...Parameters` constructors private to avoid implementors relying on them and ending up in this same situation. I wanted them package private so that they can be used internally, but `BatchedExecutionStrategy` is in a different package so there wasn't much point.
- I added Javadoc to the `@Deprecated` methods only to link to the new options, but since this is not a full Javadoc, a lot of warnings are generated during the build. Would you rather have the incomplete docs removed or maybe provide the text you'd like in the Javadoc, or would you like me to add my own?

Now, the question come back to the `TypeResolver`: do I re-introduce the Java 8 dependent `default` for backwards compatibility, or do I implement it in a different way?
One idea I have is to introduce a new interface, say `TypeDecider`, that would provide the new functionality: `GraphQLObjectType getType(TypeResolutionEnvironment env)`, and then change the builders that accept `TypeResolver`s like this:

```
//wrap the resolver in a simple TypeDecider that just delegates to it
public Builder typeResolver(TypeResolver typeResolver) {
    this.typeDecider = new SimpleTypeDecider(typeResolver);
    return this;
 }

//add new method that accepts TypeDecider directly
public Builder typeDecider(TypeDecider typeDecider) {
    this.typeDecider = typeDecider;
    return this;
 }
```